### PR TITLE
Document layers

### PR DESCRIPTION
**Pull request by graham63**

### CODE REVIEW

File: docs/layers/transform_layers.rst [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+
+________________________________________
+
+.. _Bernoulli:
+
+----------------------------------------
+Bernoulli
+----------------------------------------
+
+The Bernoulli layer is a random tensor with a Bernoulli distribution
```
Comment (timmoon10) [inline docs/layers/transform_layers.rst:81]:
```suggestion
The Bernoulli layer is a random tensor with a Bernoulli distribution.
```

File: docs/layers/transform_layers.rst [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+________________________________________
+
+
+.. _InTopK:
+
+----------------------------------------
+InTopK
+----------------------------------------
+
+The InTopK layer is a one-hot vector indicating top-k entries Output
```
Comment (timmoon10) [inline docs/layers/transform_layers.rst:328]:
```suggestion
The InTopK layer is a one-hot vector indicating top-k entries. Output
```

File: docs/layers/transform_layers.rst [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+________________________________________
+
+
+.. _Slice:
+
+----------------------------------------
+Slice
+----------------------------------------
+
+The Slice layer slices a tensor along a specified dimension The tensor
```
Comment (timmoon10) [inline docs/layers/transform_layers.rst:524]:
```suggestion
The Slice layer slices a tensor along a specified dimension. The tensor
```

File: docs/layers/learning_layers.rst [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+________________________________________
+
+
+.. _ChannelwiseFullyConnected:
+
+----------------------------------------
+ChannelwiseFullyConnected
+----------------------------------------
+
+The ChannelwiseFullyConnected layer appliesan  affine transformation
```
Comment (timmoon10) [inline docs/layers/learning_layers.rst:34]:
```suggestion
The ChannelwiseFullyConnected layer applies an affine transformation
```

File: docs/layers/learning_layers.rst [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+bias vectors :math:`a,b\in\mathbb{R}^{d_1}`:
+
+.. math::
+
+   Y_{i,j,\cdots} = a_i X_{i,j,\cdots} + b_i
+
+The scale and bias vectors are fused into a single weights tensor to
+reduce the number of gradient allreduces during backprop. In
+particular, the weights tensor is a
+:math:`\text{num\_channels} \times 2` matrix, where the first column
```
Comment (timmoon10) [inline docs/layers/learning_layers.rst:97]:
```suggestion
:math:`\text{num_channels} \times 2` matrix, where the first column
```

File: docs/layers/learning_layers.rst [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+________________________________________
+
+
+.. _Deconvolution:
+
+----------------------------------------
+Deconvolution
+----------------------------------------
+
+Deconvolution Layer. Reverses the effects of convolution.
```
Comment (timmoon10) [inline docs/layers/learning_layers.rst:235]:
```suggestion
Deconvolution Layer. Transpose of convolution.
```
Once https://github.com/LLNL/lbann/pull/1997 is merged, the text should be ported here.

File: docs/layers/learning_layers.rst [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+:math:`\text{sequence\_length}\times\text{input\_size}`) and a 2D
+initial hidden state (
+:math:`\text{num\_layers}times\text{hidden\_size}`).
+
+Uses four weights per GRU cell: "ih\_matrix" (
+:math:`3 \text{hidden\_size}\times\text{input\_size}` for layer 0 and
+:math:`3 \text{hidden\_size}\times\text{hidden\_size}` for other layers),
+"hh\_matrix" (:math:`3 \text{hidden\_size}\times\text{hidden\_size}`),
+"ih_bias" (:math:`3 \text{hidden\_size}`), "hh_bias"
+(:math:`3 \text{hidden\_size}`).
```
Comment (timmoon10) [inline docs/layers/learning_layers.rst:408]:
```suggestion
Expects two inputs: a 2D input sequence (
:math:`\text{sequence_length}\times\text{input_size}`) and a 2D
initial hidden state (
:math:`\text{num_layers}\times\text{hidden_size}`).

Uses four weights per GRU cell: "ih\_matrix" (
:math:`3 \text{hidden_size}\times\text{input_size}` for layer 0 and
:math:`3 \text{hidden_size}\times\text{hidden_size}` for other layers),
"hh\_matrix" (:math:`3 \text{hidden_size}\times\text{hidden_size}`),
"ih_bias" (:math:`3 \text{hidden_size}`), "hh_bias"
(:math:`3 \text{hidden_size}`).
```

File: docs/layers/learning_layers.rst [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+
+Support is experimental and requires either cuDNN (on GPU) or oneDNN
+(on CPU).
+
+.. todo:: Support bidirectional RNNs
```
Comment (timmoon10) [inline docs/layers/learning_layers.rst:413]: This is putting everything inside of the todo block. Maybe:
```suggestion
    .. todo:: Support bidirectional RNNs
    
Arguments:
```

File: docs/layers/math_layers.rst [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+________________________________________
+
+
+.. _DFTAbs:
+
+----------------------------------------
+DFTAbs
+----------------------------------------
+
+Absolute value of discrete Fourier transform One-, two-, or
```
Comment (timmoon10) [inline docs/layers/math_layers.rst:27]:
```suggestion
Absolute value of discrete Fourier transform. One-, two-, or
```

File: docs/layers/regularization_layers.rst [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+
+.. _Dropout:
+
+----------------------------------------
+Dropout
+----------------------------------------
+
+Probabilistically drop tensor entries
+
+The values multiplied by 1/(keep probability) at training time. Keep
```
Comment (timmoon10) [inline docs/layers/regularization_layers.rst:99]:
```suggestion
The values are multiplied by 1/(keep probability) at training time. Keep
```

**Review by timmoon10 — APPROVED** [review]
I think this is a good layout. It looks nice and I like the tables summarizing the layers in each page.

Comments:
- I'm not a fan of how we handle the one-line descriptions:
  ```
  The <name> layer is <brief description>. <detailed description>.
  ```
  It feels redundant and [Google doesn't like it](https://developers.google.com/style/api-reference-comments). Could we change it to something like:
  ```
  <brief description>
  
  <detailed description>
  ```
- When a layer has no additional arguments, I think putting `Arguments: None` is redundant.
- I notice the argument types are Protobuf types, not Python. Most of the conversions are obvious, but some of them are a bit idiosyncratic (e.g. using `google.protobuf.BoolValue` as a hack to have non-false default values for `bool`).

After fixing the typos, I think this is good to merge. I can take a followup pass to deal with the Protobuf/Python type issue.

**Review by bvanessen — APPROVED** [review]
LGTM.  I reviewed and applied @timmoon10 suggestions.

File: docs/layers.rst [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+.. _using-layers:
+
+------------------------------------------------
+Using Layers
+------------------------------------------------
+
+Layers are used by adding them to the python front end with the
+appropriate arguments and passing them as a list into the
+model. More information about LBANN's layers can be found in
+:ref:`layers-list`. See :ref:`layer-options` for layer parameter option.
```
Comment (benson31) [inline docs/layers.rst:34]: These are both dangling references.

**Review by benson31 — CHANGES_REQUESTED** [review]
Generally looks good. Just fix the dangling references and :shipit:

**Review by benson31 — APPROVED** [review]
LGTM

**Pull request merged**